prover/src/gadgets/transpose.rs

```diff
@@ -0,0 +1,103 @@
+    pub(crate) output: [Col<B32>; 16],
+    pub(crate) projected: [Col<B8>; 64],
+    pub(crate) zero_padded: [Col<B8, 4>; 64],
+    pub(crate) transposed: [Col<B8, 4>; 16],
+}
+
+impl TransposeColumns {
+    pub(crate) fn new(table: &mut TableBuilder, input: [Col<B8, 8>; 8]) -> Self {
+        // First, we project the values into independent B8 columns.
+        let projected_temp: [[Col<B8>; 8]; 8] = from_fn(|i| {
+            from_fn(|j| {
+                table.add_selected_block::<_, 8, 1>(format!("projected_{i}_{j}"), input[i], j)
+            })
+        });
+        // We take the projected values into the correct (transposed) order.
+        let projected = from_fn(|i| projected_temp[i % 8][i / 8]);
+
+        // Now, we need to construct the B32 elements we can read in memory.
+        // We zeropad the projected values to go from `Col<B8>` to `Col<B8, 4>`.
+        let zero_padded = from_fn(|i| {
+            table.add_zero_pad::<_, 1, 4>(format!("zero_padded_{i}"), projected[i], i % 4)
+        });
+        // Finally, we sum each array of B8 to get the correct B32 values.
+        let transposed: [Col<B8, 4>; 16] = zero_padded
+            .chunks(4)
+            .enumerate()
+            .map(|(i, cols)| {
+                let expr: Expr<B8, 4> = cols
+                    .iter()
+                    .map(|&col| col.into())
+                    .reduce(|acc, item| acc + item)
+                    .unwrap();
+                table.add_computed(format!("zero_padded_sums_{i}"), expr)
+            })
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+
+        let output = (0..16)
+            .map(|i| table.add_packed(format!("packed_transpose_{i}"), transposed[i]))
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+
+        Self {
+            input,
+            output,
+            projected,
+            zero_padded,
+            transposed,
+        }
+    }
+
+    pub fn populate<T>(
+        &self,
+        index: &mut TableWitnessSegment<ProverPackedField>,
+        inputs: T,
+    ) -> Result<(), anyhow::Error>
+    where
+        T: Iterator<Item = [u8; 64]>,
+    {
+        let mut input = (0..8)
+            .map(|i| index.get_mut(self.input[i]))
+            .collect::<Result<Vec<_>, _>>()?;
+        let mut projected = (0..64)
+            .map(|i| index.get_mut_as(self.projected[i]))
+            .collect::<Result<Vec<RefMut<'_, [u8]>>, _>>()?;
+        let mut zero_padded = (0..64)
+            .map(|i| index.get_mut_as(self.zero_padded[i]))
+            .collect::<Result<Vec<RefMut<'_, [[u8; 4]]>>, _>>()?;
+        let mut transposed = (0..16)
+            .map(|i| index.get_mut(self.transposed[i]))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        for (i, ev_input) in inputs.enumerate() {
+            for j in 0..8 {
+                for k in 0..8 {
+                    set_packed_slice(&mut input[j], i * 8 + k, B8::from(ev_input[k * 8 + j]));
```

Comment:
It looks like you are setting twice the inputs? because is also being done [here](https://github.com/PetraProver/PetraVM/blob/linda/add-groestl/prover/src/opcodes/groestl.rs#L336)
I think it should be only done there, and you may eliminate `inputs` from `populate`